source/features/useful-not-found-page.tsx

```diff
@@ -28,119 +25,360 @@ async function checkAnchor(anchor: HTMLAnchorElement): Promise<void> {
+	const currentBranch = parts[3];
+
+	if (!currentBranch) {
+		return;
+	}
+
+	const defaultBranch = await getDefaultBranch();
+	if (currentBranch === defaultBranch) {
+		return;
+	}
+
+	// Change branch
+	parts[3] = defaultBranch;
+	const url = "/" + parts.join("/");
+	// Check if that path exists
 	if (await is404(location.origin + url)) {
 		return;
 	}
 
-	bar.after(
-		<p className="container mt-4 text-center">
-			See also the {getType()}’s <a href={url}>commit history</a>.
-		</p>,
+	// If it does, return it
+	return url;
+}
+
+async function getLastCommitForFile(
+	branch?: string
+): Promise<Record<string, string | undefined> | undefined> {
+	// Get the file path from the parts
+	const parts = parseCurrentURL();
+	let filePath = parts.slice(4).join("/");
+
+	// Get the last 2 commits that include the file
+	const commitsForFileResponse = await api.v4(`
+		repository() {
+			ref(qualifiedName: "${branch ?? parts[3]}") {
+				target {
+					... on Commit {
+						history(first: 2, path: "${filePath}") {
+							nodes {
+								oid
+							}
+						}
+					}
+				}
+			}
+		}
+	`);
+	// The latest commit will be the first object in the array
+	const previousCommitsIncludingFile =
+		commitsForFileResponse.repository.ref.target.history.nodes[0];
+	if (!previousCommitsIncludingFile) {
+		return;
+	}
+
+	// Get a list of changes that happened in the repo with this commit
+
+	// API v4 doesn't support retrieving a list of changed files for a commit:
+	// https://github.community/t/graphql-api-get-list-of-files-related-to-commit/14047/2
+	const lastCommitInfo = await api.v3(
+		`commits/${previousCommitsIncludingFile.oid as string}`
 	);
+
+	// Check what happened to this particular file
+	const fileInfo = lastCommitInfo.files.find((file: Record<string, string>) => {
+		if (file.filename === filePath) {
+			return true;
+		}
+
+		if (file.status === "renamed" && file.previous_filename === filePath) {
+			return true;
+		}
+
+		return false;
+	});
+	if (!fileInfo) {
+		return;
+	}
+
+	// Get info
+	const commitSha = lastCommitInfo.sha.slice(0, 8);
+	const commitDate = lastCommitInfo.commit.committer.date;
+	const linkToCommit = lastCommitInfo.html_url;
+
+	parts[2] = "commits";
+	const linkToCommitHistory = "https://github.com/" + parts.join("/");
+
+	if (fileInfo.status === "removed") {
+		const linkToLastVersion = fileInfo.blob_url;
+
+		return {
+			type: "removed",
+			filePath,
+			commitSha,
+			commitDate,
+			linkToCommit,
+			linkToLastVersion,
+			linkToCommitHistory,
+		};
+	}
+
+	if (fileInfo.status === "renamed") {
+		filePath = fileInfo.previous_filename || filePath;
+		const newFilePath = fileInfo.previous_filename ? filePath : undefined;
+
+		const linkToNewVersion = fileInfo.blob_url;
+		let linkToLastVersion = "";
+
+		const lastToLastCommit =
+			commitsForFileResponse.repository.ref.target.history.nodes[1];
+		if (lastToLastCommit) {
+			// HACK: Generate a blob URL for the file at the time of the previous commit
```

Comment:
Try using GitHubURL to generate/change the URL instead. Starting from the current page, you should be able to just set 2 parts (I don't remember if these are the correct props):

```js
url.assign({
	ref: lastToLastCommit.oid,
	filePath
})
```